spec/classes/repo_spec.rb

```diff
@@ -3,24 +3,50 @@
 describe 'candlepin::repo' do
   on_supported_os.each do |os, os_facts|
     context "on #{os}" do
-      let(:facts) { os_facts }
-      let(:params) do
-        { version: 'nightly' }
-      end
+      describe 'with nightly version' do
+        let(:facts) { os_facts }
+        let(:params) do
+          { version: 'nightly' }
+        end
+
+        it { is_expected.to compile.with_all_deps }
 
-      it { is_expected.to compile.with_all_deps }
+        it do
+          is_expected.to contain_yumrepo('candlepin')
+            .with_descr('Candlepin: an open source entitlement management system.')
+            .with_baseurl("https://yum.theforeman.org/katello/nightly/candlepin/el#{facts[:os]['release']['major']}/\$basearch/")
+            .with_gpgkey(nil)
+            .with_gpgcheck(false)
+            .with_enabled(true)
+            .that_comes_before('Anchor[candlepin::repo]')
+        end
 
-      it do
-        is_expected.to contain_yumrepo('candlepin')
-          .with_descr('Candlepin: an open source entitlement management system.')
-          .with_baseurl("https://yum.theforeman.org/katello/nightly/candlepin/el#{facts[:os]['release']['major']}/\$basearch/")
-          .with_gpgkey(nil)
-          .with_gpgcheck(false)
-          .with_enabled(true)
-          .that_comes_before('Anchor[candlepin::repo]')
+        it { is_expected.to contain_anchor('candlepin::repo') }
       end
 
-      it { is_expected.to contain_anchor('candlepin::repo') }
+      describe 'with baseurl' do
+        let(:facts) { os_facts }
```

Comment:
And if you set `facts` at the context level you can leave it out here.
```suggestion
```

Here you can even apply the nesting I talked about. For readability I left out some assertions, but it comes down to:
```ruby
require 'spec_helper'

describe 'candlepin::repo' do
  on_supported_os.each do |os, os_facts|
    context "on #{os}" do
      let(:facts) { os_facts } 

      describe 'with nightly version' do
        let(:params) do
          { version: 'nightly' } 
        end

        it { is_expected.to compile.with_all_deps } 

        describe 'with baseurl' do
          let(:params) { super().merge(baseurl: "https://stagingyum.theforeman.org/candlepin/nightly/el#{facts[:os]['release']['major']}/\$basearch/") } 

          it { is_expected.to compile.with_all_deps } 
        end
      end
    end
  end
end
```

You can also see it in the output:
```console
$ be rspec --format documentation spec/classes/repo_spec.rb 

candlepin::repo
  on centos-8-x86_64
    with nightly version
      is expected to compile into a catalogue without dependency cycles
      is expected to contain Yumrepo[candlepin] that comes before Anchor[candlepin::repo]
      is expected to contain Anchor[candlepin::repo]
      with baseurl
        is expected to compile into a catalogue without dependency cycles
        is expected to contain Yumrepo[candlepin] that comes before Anchor[candlepin::repo]
        is expected to contain Anchor[candlepin::repo]
  on redhat-8-x86_64
    with nightly version
      is expected to compile into a catalogue without dependency cycles
      is expected to contain Yumrepo[candlepin] that comes before Anchor[candlepin::repo]
      is expected to contain Anchor[candlepin::repo]
      with baseurl
        is expected to compile into a catalogue without dependency cycles
        is expected to contain Yumrepo[candlepin] that comes before Anchor[candlepin::repo]
        is expected to contain Anchor[candlepin::repo]
```

And in failures too:
```console
$ be rspec spec/classes/repo_spec.rb 
...F.....F...

Failures:

  1) candlepin::repo on centos-8-x86_64 with nightly version with baseurl is expected not to compile into a catalogue without dependency cycles
     Failure/Error: it { is_expected.not_to compile.with_all_deps }
       expected that the catalogue would not compile but it does
     # ./spec/classes/repo_spec.rb:30:in `block (6 levels) in <top (required)>'
     # /usr/share/gems/gems/bundler-2.4.10/libexec/bundle:45:in `block in <top (required)>'
     # /usr/share/gems/gems/bundler-2.4.10/libexec/bundle:33:in `<top (required)>'
```

Note how it combined the various descriptions and contexts. Technically I think `describe` and `context` do the same thing and are only separated out with the intention that a `describe` actively tests something and `context` only sets up a context.